helix-core/src/main/java/org/apache/helix/controller/stages/PerReplicaThrottleStage.java

```diff
@@ -220,6 +271,224 @@ private MessageOutput throttlePerReplicaMessages(IdealState idealState,
+      Map<Partition, Map<String, Integer>> currentStateCountsByPartition) {
+    String resourceName = idealState.getResourceName();
+    Map<String, List<String>> preferenceLists =
+        bestPossibleStateOutput.getPreferenceLists(resourceName);
+    for (Partition partition : selectedResourceMessages.keySet()) {
+      Map<String, String> currentStateMap =
+          currentStateOutput.getCurrentStateMap(resourceName, partition);
+
+      Map<String, Integer> expectedStateCounts = new HashMap<>();
+      Map<String, Integer> currentStateCounts = new HashMap<>();
+      getPartitionExpectedAndCurrentStateCountMap(partition, preferenceLists, idealState, cache,
+          currentStateMap, expectedStateCounts, currentStateCounts);
+
+      // save these two maps for later usage
+      expectedStateCountByPartition.put(partition, expectedStateCounts);
+      currentStateCountsByPartition.put(partition, currentStateCounts);
+    }
+  }
+
+  /*
+   * Classify the messages of each partition into recovery and load messages.
+   */
+  private void classifyMessages(String resourceName, IdealState idealState,
+      ResourceControllerDataProvider cache, Map<Partition, List<Message>> selectedResourceMessages,
+      List<Message> recoveryMessages, List<Message> loadMessages,
+      Map<Partition, Map<String, Integer>> expectedStateCountByPartition,
+      Map<Partition, Map<String, Integer>> currentStateCountsByPartition) {
+    LogUtil.logInfo(logger, _eventId,
+        String.format("Classify message for resource: %s", resourceName));
+
+    for (Partition partition : selectedResourceMessages.keySet()) {
+      Map<String, Integer> expectedStateCountMap = expectedStateCountByPartition.get(partition);
+      Map<String, Integer> currentStateCounts = currentStateCountsByPartition.get(partition);
+
+      List<Message> partitionMessages = selectedResourceMessages.get(partition);
+      if (partitionMessages == null) {
+        continue;
+      }
+
+      String stateModelDefName = idealState.getStateModelDefRef();
+      StateModelDefinition stateModelDef = cache.getStateModelDef(stateModelDefName);
+      // sort partitionMessages based on transition priority and then creation timestamp for transition message
+      // TODO: sort messages in same partition in next PR
+      Set<String> disabledInstances =
+          cache.getDisabledInstancesForPartition(resourceName, partition.getPartitionName());
+      for (Message msg : partitionMessages) {
+        if (!Message.MessageType.STATE_TRANSITION.name().equals(msg.getMsgType())) {
+          if (logger.isDebugEnabled()) {
+            LogUtil.logDebug(logger, _eventId, String
+                .format("Message: %s not subject to throttle in resource: %s with type %s", msg,
+                    resourceName, msg.getMsgType()));
+          }
+          continue;
+        }
+
+        boolean isUpward = !isDownwardTransition(idealState, cache, msg);
```

Comment:
Hasn't we decided to put only "basic throttle" logic in this PR? The DownwardTransition shall be put in a later PR, right?

Reply:
removed. special treatment of like this would be in next PRs

Reply:
I was looking back to this previous comment because I remember that I suggest that we put only "basic throttle" logic in this PR. But it seems not the case : )

Reply:
I did removed dropped partition and error partition handling.

 Removed one more part
 ```
      Set<String> disabledInstances =
          cache.getDisabledInstancesForPartition(resourceName, partition.getPartitionName());
```
So the code is now like this:
```
      for (Message msg : partitionMessages) {
        if (!Message.MessageType.STATE_TRANSITION.name().equals(msg.getMsgType())) {
          logger.debug(
              "Event info: {} Message: {} not subject to throttle in resource: {} with type {}",
              _eventId, msg, resourceName, msg.getMsgType());
          continue;
        }

        boolean isUpward = !isDownwardTransition(stateModelDef, msg);

        // TODO: add disabled disabled instance special treatment

        String toState = msg.getToState();

        // TODO: dropped and Error state special treatment

        Integer minimumRequiredCount = partitionExpectedStateCounts.getOrDefault(toState, 0);
        Integer currentCount = partitionCurrentStateCounts.getOrDefault(toState, 0);

        //
        if (isUpward && (currentCount < minimumRequiredCount)) {
          recoveryMessages.add(msg);
          // It is critical to increase toState value by one here. For example, current state
          // of 3 replica in a partition is (M, O, O). Two messages here bringing up the two O to S.
          // In this case, the first O->S would be classified as recovery. Then this line would
          // increase the S state in partitionCurrentStateCounts value by one. Next O->S message
          // would be correctly marked as load message.
          partitionCurrentStateCounts.put(toState, currentCount + 1);
        } else {
          loadMessages.add(msg);
        }
      }

```

Let me know if there is anything else? Or we can comment on the new code.